utils/zerovec/src/varzerovec/owned.rs

```diff
@@ -115,6 +167,118 @@ impl<T: AsVarULE> VarZeroVecOwned<T> {
         &self.entire_slice
     }
 
+    // Invalidate and resize the data at an index, optionally inserting or removing the index.
+    // Also updates affected indices and the length.
+    // Returns a slice to the new element data - it doesn't contain uninitialized data but its value is indeterminate.
+    unsafe fn shift(&mut self, index: usize, new_size: u32, shift_type: ShiftType) -> &mut [u8] {
+        // The format of the encoded data is:
+        //  - four bytes of "len"
+        //  - len*4 bytes for an array of indices
+        //  - the actual data to which the indices point
+        //
+        // When inserting or removing an element, the size of the indices segment must be changed,
+        // so the data before the target element must be shifted by 4 bytes in addition to the
+        // shifting needed for the new element size.
+        let len = self.len();
+        let slice_len = self.entire_slice.len();
+
+        let prev_element = match shift_type {
+            ShiftType::Insert => {
+                let pos = self.element_position_unchecked(index);
+                // In the case of an insert, there's no previous element,
+                // so it's an empty range at the new position.
+                pos..pos
+            }
+            _ => self.element_range_unchecked(index),
+        };
+
+        // How much shifting must be done due to removal/insertion of an index.
+        let index_shift: i64 = match shift_type {
+            ShiftType::Insert => 4,
+            ShiftType::Replace => 0,
+            ShiftType::Remove => -4,
+        };
+        // The total shift in byte size of the owned slice.
+        let shift: i64 =
+            new_size as i64 - (prev_element.end - prev_element.start) as i64 + index_shift;
+        let new_slice_len = slice_len.wrapping_add(shift as usize);
+        if shift > 0 {
+            if new_slice_len > u32::MAX as usize {
+                panic!(
+                    "Attempted to grow VarZeroVec to an encoded size that does not fit within a u32"
+                );
+            }
+            self.entire_slice.reserve(shift as usize);
+        }
+
+        // Now that we've ensured there's enough space, we can shift the data around.
+        let first_affected_index = {
+            // Note: There are no references introduced between pointer creation and pointer use, and all
+            //       raw pointers are derived from a single &mut. This preserves pointer provenance.
+            let slice_range = self.entire_slice.as_mut_ptr_range();
+            let data_start = slice_range.start.add(4 + len * 4);
+            let prev_element_p =
+                data_start.add(prev_element.start)..data_start.add(prev_element.end);
+
+            let index_range = {
```

Comment:
nit: document what this variable is for, it seems to be used a bit differently

Reply:
Done, how's that?

Reply:
(are you missing some git pushes?)

Reply:
Honestly I was hoping you wouldn't be up and I forgot this doesn't send all the comments in a single batch

Reply:
heh that works too :smile: 

I usually wouldn't be but i got bored and was chatting with eddyb about some gnarly trait stuff